make torch.histc ignore NaNs on CPU

yuguo68 opened the pull request:
Summary: cuda torch.histc already ignores NaNs

Test Plan: unittest added

Differential Revision: D39911272

fix https://github.com/pytorch/pytorch/issues/85853

mruberry commented on aten/src/ATen/native/cpu/HistogramKernel.cpp:
```diff
@@ -149,7 +149,7 @@ void histogramdd_cpu_contiguous(Tensor& hist, const TensorList& bin_edges,
 
                 // Skips elements which fall outside the specified bins
-                if (elt < leftmost_edge[dim] || rightmost_edge[dim] < elt) {
+                if (!(elt >= leftmost_edge[dim] && elt <= rightmost_edge[dim])) {
```
This seems OK but let's extend the above comment that this check skips elements that fall outside the bins AND NaNs

mruberry commented on test/test_reductions.py:
```diff
@@ -2795,9 +2795,14 @@ def test_histc(self, device):
+        # tensor with nan; min, max not provided -- should throw a RuntimeError
         with self.assertRaisesRegex(RuntimeError, r'range of \[nan, nan\] is not finite'):
             torch.histc(torch.tensor([float("nan")], dtype=torch.float, device=device))
+        # tensor with nan; min, max provided
```
Nice test extension -- let's update this comment to explain that NaNs should not be counted in the output

mruberry commented:
Looks good overall, @yuguo68!

This PR should also update the documentation for histc here:

https://github.com/pytorch/pytorch/blob/dc63948dc9dbe4c224a78a7c14f406893f6fd381/torch/_torch_docs.py#L4985

to explain that NaN elements are also ignored.

yuguo68 commented:
> Looks good overall, @yuguo68!
> 
> This PR should also update the documentation for histc here:
> 
> https://github.com/pytorch/pytorch/blob/dc63948dc9dbe4c224a78a7c14f406893f6fd381/torch/_torch_docs.py#L4985
> 
> to explain that NaN elements are also ignored.

thank you for reviewing! Will add the comments in code/doc shortly.

yuguo68 committed:
make torch.histc ignore NaNs on CPU (#85870)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85870

cuda torch.histc already ignores NaNs

Test Plan: unittest added

Differential Revision: D39911272

fbshipit-source-id: 22f5c5c4ba98dafa0e1fbc2294b95af134ad3d64

yuguo68 commented:
test failure `TestCommonCUDA.test_python_ref_executor__refs_tanh_executor_aten_cuda_complex128` is not related

ngimel commented:
@pytorchbot rebase && merge -f "test failure urelated"

ngimel commented:
@pytorchbot --help

ngimel commented:
@pytorchbot rebase

yuguo68 commented:
@mruberry could you review this PR again? thanks!

ngimel commented:
@pytorchbot merge -g

yuguo68 commented:
> @pytorchbot merge -g

thank you! thought we need all code owners to approve.
